server/rating_service/rating_service.py

```diff
@@ -0,0 +1,365 @@
+                            game_player_stats.c.playerId == player_id,
+                            game_player_stats.c.mean < game_player_stats.c.after_mean,
+                        )
+                    )
+                )
+                result_proxy = await conn.execute(count_wins_query)
+                won_games = await result_proxy.scalar()
+            else:
+                won_games = row["winGames"]
+
+            insertion_sql = leaderboard_rating.insert().values(
+                login_id=player_id,
+                mean=row["mean"],
+                deviation=row["deviation"],
+                total_games=row["numGames"],
+                won_games=won_games,
+                leaderboard_id=self._rating_type_ids[rating_type.value],
+            )
+            await conn.execute(insertion_sql)
+
+            return Rating(row["mean"], row["deviation"])
+
+    async def _create_default_rating(
+        self, conn, player_id: int, rating_type: RatingType
+    ):
+        default_mean = config.START_RATING_MEAN
+        default_deviation = config.START_RATING_DEV
+        rating_type_id = self._rating_type_ids.get(rating_type.value)
+
+        insertion_sql = leaderboard_rating.insert().values(
+            login_id=player_id,
+            mean=default_mean,
+            deviation=default_deviation,
+            total_games=0,
+            won_games=0,
+            leaderboard_id=rating_type_id,
+        )
+        await conn.execute(insertion_sql)
+
+        return Rating(default_mean, default_deviation)
+
+    async def _persist_rating_changes(
+        self,
+        game_id: int,
+        rating_type: RatingType,
+        old_ratings: Dict[PlayerID, Rating],
+        new_ratings: Dict[PlayerID, Rating],
+        outcomes: Dict[PlayerID, GameOutcome],
+    ) -> None:
+        """
+        Persist computed ratings to the respective players' selected rating
+        """
+        self._logger.info("Saving rating change stats for game %i", game_id)
+
+        async with self._db.acquire() as conn:
+            for player_id, new_rating in new_ratings.items():
+                old_rating = old_ratings[player_id]
+                self._logger.debug(
+                    "New %s rating for player with id %s: %s -> %s",
+                    rating_type,
+                    player_id,
+                    old_rating,
+                    new_rating,
+                )
+
+                gps_update_sql = (
+                    game_player_stats.update()
+                    .where(
+                        and_(
+                            game_player_stats.c.playerId == player_id,
+                            game_player_stats.c.gameId == game_id,
+                        )
+                    )
+                    .values(
+                        after_mean=new_rating.mu,
+                        after_deviation=new_rating.sigma,
+                        mean=old_rating.mu,
+                        deviation=old_rating.sigma,
```

Comment:
This now becomes duplicated across both `game_player_stats` and `leaderboard_rating_journal`. I think in a perfect world these fields would be removed from `game_player_stats` and instead rating change history would only be recorded in the journal. But I'd like to know what @Brutus5000 thinks about this.

If we drop those columns from `game_player_stats` we'll have to back-fill `leaderboard_rating_journal` first in order to preserve the history, which again is going to be a lot of work. Plus apps that display your rating history (like the client) will need to be changed to use the journal table, although this should be pretty trivial (simply calling a different API endpoint).

Reply:
Back-filling `leaderboard_rating_journal` would make the global/ladder game count and win ratio queries discussed above significantly simpler, I think.

Reply:
Dropping these fields from `game_player_stats` would be the plan, but that is a backward-compatible change and will break all clients touching these.
We might need an API workaround to still populate them :thinking: